core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/context/internal/CarbonContextDataHolder.java

```diff
@@ -1287,7 +1287,7 @@ public void startTenantFlow() {
      */
     public void endTenantFlow() {
         Stack<CarbonContextDataHolder> carbonContextDataHolders = parentContextHolderStack.get();
-        if (carbonContextDataHolders != null) {
+        if (carbonContextDataHolders != null && !carbonContextDataHolders.isEmpty()) {
```

Comment:
We need to find the root cause on why the stack becomes empty rather than checking for isEmpty. This is discussed in the thread - "[Dev] [ESB] EmptyStackException when resuming a paused message	processor"

https://www.mail-archive.com/dev@wso2.org/msg47287.html